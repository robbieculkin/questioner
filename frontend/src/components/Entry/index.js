import React, { Component } from 'react';
import './index.scss';

class Entry extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: ''
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.handleTextEntry = this.handleTextEntry.bind(this);
  }

  handleChange(event) {
    this.setState({ value: event.target.value });
  }

  handleKeyPress(event) {
    // Submit textentry when <ENTER> is pressed (instead of newline)
    const { value } = this.state;

    if (event.key === 'Enter') {
      event.preventDefault();
      this.handleTextEntry(value);
    }
  }

  handleTextEntry(value) {
    const { onTextEntry } = this.props;

    onTextEntry(value);
    this.setState({ value: '' });
  }

  render() {
    const { value } = this.state;

    return (
      <div className='entry'>
        <textarea
          id='entryField'
          type='text'
          value={value}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
        />
        <button onClick={() => this.handleTextEntry(value)}>Next</button>
      </div>
    );
  }
}

export default Entry;