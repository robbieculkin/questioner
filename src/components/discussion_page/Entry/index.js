import React, { Component } from 'react';
import PropTypes from 'prop-types';
import getCaretCoordinates from 'textarea-caret';
import './index.scss';

const getCaretPosition = () => {
  // Return caret position measured as a character count
  // Return -1 if there is a selection
  const elt = document.getElementById('entryField');
  const start = elt.selectionStart;
  const end = elt.selectionEnd;

  return start === end ? end : -1;
};

const getCaretStyle = unfocused => {
  // visibility hidden if argument provided/falsy
  const caretPosition = getCaretPosition();
  const elt = document.getElementById('entryField');

  if (!unfocused && caretPosition >= 0) {
    const caret = getCaretCoordinates(elt, caretPosition);
    const rect = elt.getBoundingClientRect();

    return {
      top: rect.top + caret.top,
      left: rect.left + caret.left,
      visibility: 'visible'
    };
  } else {
    return {
      visibility: 'hidden'
    };
  }
};

class Entry extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: '',
      caretStyle: {
        visibility: 'hidden'
      }
    };

    this.handleClick = this.handleClick.bind(this);
    this.handleScroll = this.handleScroll.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.handleTextEntry = this.handleTextEntry.bind(this);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll);
    this.handleBlur();
    setTimeout(this.handleFocus, 100);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll);
  }

  handleClick() {
    this.setState({ caretStyle: getCaretStyle() });
  }

  handleScroll() {
    this.setState({ caretStyle: getCaretStyle() });
  }

  handleFocus() {
    this.setState({ caretStyle: getCaretStyle() });
  }

  handleBlur() {
    this.setState({ caretStyle: getCaretStyle('unfocused') });
  }

  handleChange(event) {
    this.setState({
      value: event.target.value,
    });
  }

  handleKeyDown(event) {
    // Submit textentry when <ENTER> is pressed (instead of newline)
    const { value } = this.state;
    this.setState({ caretStyle: getCaretStyle() });

    if (event.key === 'Enter') {
      event.preventDefault();
      this.handleTextEntry(value);
    }
  }

  handleKeyUp() {
    this.setState({ caretStyle: getCaretStyle() });
  }

  handleTextEntry(value) {
    const { onTextEntry } = this.props;
    if (value === '')
      return;

    onTextEntry(value);
    this.setState({
      value: '',
      caretStyle: getCaretStyle()
    });
  }

  render() {
    const { value, caretStyle } = this.state;

    return (
      <div className='entry'>
        <textarea
          id='entryField'
          type='text'
          autoFocus
          value={value}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          onKeyUp={this.handleKeyUp}
          onClick={this.handleClick}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
        <div className='caret' style={caretStyle}></div>
      </div>
    );
  }
}

Entry.propTypes = {
  onTextEntry: PropTypes.func
};

export default Entry;