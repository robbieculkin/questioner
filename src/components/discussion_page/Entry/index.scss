.entry {
    width: 100%;

    textarea {
        /* Fix issues with Grammarly extension! */
        /* Write JS to resize textarea when too many lines */
        display: block;
        min-height: 50px;
        width: 100%;
        margin-bottom: 10px;
        border: none;
        resize: none;
        overflow: hidden;
        outline: none;
        font-size: 24px;
        background: #eee;
    }

    button {
        width: 100px;
        height: 40px;
        outline: none;
        border: 1px solid #aaaaaa;
        border-radius: 0;
        font-size: 20px;
    }
}