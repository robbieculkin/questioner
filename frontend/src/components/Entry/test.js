import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Entry from '.';

describe('Projects', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Entry />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Entry />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
