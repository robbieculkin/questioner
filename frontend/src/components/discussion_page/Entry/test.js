import React from 'react';
import { shallow, mount } from 'enzyme';
import Entry from '.';

describe('Entry', () => {
  it('should render correctly with no props', () => {
    const component = shallow(<Entry />);
    expect(component).toMatchSnapshot();
  });

  it('should call props fn on button click', () => {
    const onTextEntryFn = jest.fn();
    const component = shallow(<Entry onTextEntry={onTextEntryFn} />);

    component.setState({ value: 'test' });
    component
      .find('button')
      .simulate('click');

    expect(onTextEntryFn).toHaveBeenCalled();
  });

  it('should call props fn on pressing <ENTER>', () => {
    const onTextEntryFn = jest.fn();
    const component = mount(<Entry onTextEntry={onTextEntryFn} />);

    component.setState({ value: 'test' });
    component
      .find('textarea')
      .simulate('keypress', { key: 'Enter' });

    expect(component).toMatchSnapshot();
    expect(onTextEntryFn).toHaveBeenCalled();
    component.unmount();
  });
});
