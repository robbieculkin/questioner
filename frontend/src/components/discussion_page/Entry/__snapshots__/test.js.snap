// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry should call props fn on pressing <ENTER> 1`] = `
<Entry
  onTextEntry={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="entry"
  >
    <textarea
      autoFocus={true}
      id="entryField"
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</Entry>
`;

exports[`Entry should render correctly with no props 1`] = `
<div
  className="entry"
>
  <textarea
    autoFocus={true}
    id="entryField"
    onChange={[Function]}
    onKeyPress={[Function]}
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Next
  </button>
</div>
`;
